File: 3583-count-special-triplets/3583-count-special-triplets.py
class Solution:
    def specialTriplets(self, nums: List[int]) -> int:
        right = {}
        left = {}
        n = len(nums)

        for i in range(2, n):
            right[nums[i]] = right.get(nums[i], 0) + 1

        left[nums[0]] = 1

        cnt = 0
        mod = int(1e9 + 7)

        for i in range(1, n - 1):
            l = 0
            d = nums[i] * 2
            if d in left and left[d] > 0:
                l = left[d]
            r = 0
            if d in right and right[d] > 0:
                r = right[d]

            cnt = cnt + ((l * r) % mod) % mod
            right[nums[i + 1]] = right[nums[i + 1]] - 1
            left[nums[i]] = left.get(nums[i], 0) + 1

        return int(cnt % mod)

File: 3583-count-special-triplets/3583-count-special-triplets.cpp
class Solution {
public:
    // nums[i] == nums[j] * 2 == nums[k]
    int specialTriplets(vector<int>& nums) {
        unordered_map<int, int> right;
        int n = nums.size();

        for (int i = 2; i < n; i++) {
            right[nums[i]]++;
        }

        unordered_map<int, int> left;
        left[nums[0]]++;
        long long cnt = 0;
        int mod = 1e9 + 7;

        for (int i = 1; i < n - 1; i++) {
            long long l = 0;
            int d = 2 * nums[i];
            if (left.find(d) != left.end()) {
                l = left[d];
            }
            long long r = 0;
            if (right.find(d) != right.end()) {
                r = right[d];
            }
            cnt = (cnt + (l * r) % mod) % mod;
            right[nums[i + 1]]--;
            left[nums[i]]++;
        }

        return int(cnt % mod);
    }
};